### ISSUE TITLE
Add support for configuration from a distributed key-value store

### ISSUE THREAD

**Issue by guilhem** [h2o/h2o#541]
H2O is really lovely to use and configure.
But what can make it really powerful to maintain and deploy is to be able to use a distributed key/value store like `etcd` or `consul`.

I mainly have `etcd` in mind who is embedded in `coreos` and really easy to use, but it's not the only solution.

I think H2O configuration is already "k/v" ready and can be really easily put in a distributed k/v store.

Optionally we can also put a trigger to reload service on configuration change (with a "watch").

Something helpful to configure service would be some environment variables to configure service. For example:

`H2O_CONFIGURATION_STORE` (to configure file or etcd, this can be an URI)
`H2O_ETCD_USER`
etc

What are you thinking about it?

**Comment by jbergstroem** [comment]
Seeing how h2o reads yaml; couldn't you use a tool similar to [etcdtool](https://github.com/mickep76/etcdtool)?

**Comment by guilhem** [comment]
@jbergstroem yes it's always possible, but it means to have a multi process container (which is bad) or a pod with an inter-container signal send to reload configuration (also bad).
Look at _traefik_ server to see how a software can use a distributed key-value store

**Comment by yannick** [comment]
completely agree with @guilhem , its the same reason we built in kafka logging instead of piping it to a 2nd process. 
i think ideally mruby would get full access to manipulate and reload all config params and then that config management could be shipped as mruby files. 
the same goes for e.g. ACME (let's encrypt) support.

**Comment by jbergstroem** [comment]
> @guilhem said:
> @jbergstroem yes it's always possible, but it means to have a multi process container (which is bad) or a pod with an inter-container signal send to reload configuration (also bad).

Agreed on both accounts; just "getting it working".

> @yannick said:
> i think ideally mruby would get full access to manipulate and reload all config params and then that config management could be shipped as mruby files.

I like this approach too.
